ENH: Add isnat function and make comparison tests NAT specific

seberg opened the pull request:
Continuation of the discussion in gh-8413. First commit just adds a new `isnat` function, the second commit tries to clean up the warning filter in the comparison function.
Cleaning those up may be noticed by downstream projects (I am not sure). Backporting that part would be possible by implementing a local python implemented `_isnat` function instead (which is rather simple).

TODO:
- [x] I think the `assert_equal` function may say that a datetime NaT and a timedelta NaT are equal.

seberg committed:
ENH: Add isnat function

seberg committed:
BUG: Remove warning filters from comparison assert functions

These warning filters are not threadsafe, while it is nicer if the
comparison functions can be called in a threaded environment since
they do no explicite warning checking.

Less filters are also cleaner in general, though may also mean that
downstream projects see warnings that were previously hidden.

seberg commented:
There is an old PR (gh-7856), which implements `isfinite` for all types including datetime, wondering whether we should do that instead, as well or not.

charris commented on numpy/testing/utils.py:
```diff
@@ -395,13 +396,29 @@ def assert_equal(actual,desired,err_msg='',verbose=True):
+        if isnat(desired) or isnat(actual):
+            isdesnat = isnat(desired)
+            isactnat = isnat(actual)
+            if isdesnat or isactnat:
```
Could just do xor here: `if isnat(desired) ^ isnat(actual):`.

eric-wieser replied:
Or possibly more clearly, `isdesnat != isactnat`

charris commented:
> I think the assert_equal function may say that a datetime NaT and a timedelta NaT are equal.

Looks like that. I think we need to distinguish the two.

charris commented:
Otherwise, LGTM.

charris commented:
Note that this function could be implemented in pure python using views and integer comparisons. That might be an easier way to go. Same with the isfinite function, although the latter might be faster as a ufunc.

seberg commented:
For `isnat`, agree we could make it a python function. The question I guess is whether having it a true ufunc is a good consistency gain. For `isfininte`, I would say its a no-go, since it already is a ufunc.

homu commented:
:umbrella: The latest upstream changes (presumably #8452) made this pull request unmergeable. Please resolve the merge conflicts.

charris commented:
@seberg Needs rebase

seberg commented:
Rebased, and fixed the datetime vs. timedelta stuff, still needs a few tests for the array case, maybe will add them in a bit.

seberg commented:
Hmm, just noticed, you had somehow two authors, or author+commiter. Not sure how to do that, though it would seem right to add you for the second commit.

eric-wieser commented:
@seberg: wrong Pr?

seberg commented:
haha, yeah, must have mixed up some browser tabs.

charris commented:
restart tests.

charris commented:
@seberg The test for complex equality is failing. I'm going to kick this on to 1.14

charris commented:
Oops, need this for the release to fix thread safety of comparison functions.

charris commented:
@seberg Nag, nag. Consider yourself nagged ;)

seberg commented:
Yeah... well don't try funny refactoring then. We could still also do the private python version of isnat though.

charris commented on numpy/core/tests/test_datetime.py:
```diff
@@ -1918,6 +1918,36 @@ def test_datetime_y2038(self):
+        for t in np.typecodes["All"]:
+            if t in np.typecodes["Datetime"]:
+                continue
+
```
I'd drop the blank line, but not worth fixing here.

charris commented on numpy/testing/tests/test_utils.py:
```diff
@@ -144,7 +145,10 @@ def test_recarrays(self):
         c['floupa'] = a['floupa'].copy()
 
-        self._test_not_equal(c, b)
+        with suppress_warnings() as sup:
```
Unrelated fixup?

charris replied:
Oh, need to get rid of `suppress_warnings`.

seberg replied:
The safe comparison/suppression in the test filtered a bit more then just the NaT stuff (actually may have been there originally not for NaT, but for `==` operator warnings), so (not actually remembering it) removing it, the tests need to filter these warnings explicitly now. Not sure why this is the only one specifically though.

charris replied:
So this might cause some problems downstream? I'm not terribly concerned, the only way to find out is to try.

charris replied:
Hmm, looks like the `FutureWarning` will be gone in the 1.14 release.

seberg replied:
It might cause problems, but maybe its also good, since maybe some project might run into the future warning and did not notice because of this....

charris commented on numpy/testing/utils.py:
```diff
@@ -396,14 +397,24 @@ def assert_equal(actual,desired,err_msg='',verbose=True):
         pass
 
-    # Explicitly use __eq__ for comparison, ticket #2552
-    with suppress_warnings() as sup:
```
So this is the fix for thread safety?

seberg replied:
Yeah

charris commented on numpy/testing/utils.py:
```diff
@@ -396,14 +397,24 @@ def assert_equal(actual,desired,err_msg='',verbose=True):
+    except (TypeError, ValueError, NotImplementedError):
+        pass
+
+    # Explicitly use __eq__ for comparison, ticket #2552
```
?

seberg replied:
Dunno the comment was already there, probably ancient, maybe it was once `np.equal` or so.

charris commented on numpy/testing/utils.py:
```diff
@@ -674,33 +685,12 @@ def assert_array_compare(comparison, x, y, err_msg='', verbose=True,
     x = array(x, copy=False, subok=True)
     y = array(y, copy=False, subok=True)
 
-    def safe_comparison(*args, **kwargs):
```
No longer needed? What replaces this?

charris replied:
Only needed for time comparisons then?

seberg replied:
Well, it would filter a bit more (actuallylikely was first put here by Nathaniel or so for the `==` deprecations), but it seems the only test (after supporting NaT explicitely) that needed extra handling was this one test you comment on before. Not sure, maybe some of the deprecations are finished by now anyway?
Of course this might effect downstream a little here, which is why we did not go this way before after all.

seberg replied:
Heh, I am wondering if it should be worrisome that only one test needed fixup.... but my brain refuses to really think about it right now ;)

charris commented:
Looks good to me, just a few questions for clarification. Needs a 1.13 release note.

seberg commented:
Hmmm, I just found another one of these in the masked array testutils. But they don't seem to support NaNs either, so not sure I shouldn't just move the warning suppression to the affected test. Or otherwise call the `assert_equal` from here, which figures NaNs and NaTs are equal.

charris commented:
> There is an old PR (gh-7856), which implements isfinite for all types including datetime,

Could probably implement that as a define to the relevant `isnat` after this.

charris commented:
Still needs a release note.

seberg committed:
BUG: Remove warning NaT filter from masked array test utils

This may show warnings in downstream projects, which should likely
not be there. We could also drop in the normal assert_equal there,
which would support NaT, and NaN, but checks for scalarness, which
this currently does not (which also could create problems).

Works around the only test with this problem, by using the normal
assert_equal there.

seberg committed:
DOC: Document isnat addition and related warning changes

seberg commented:
Added some notes. I removed the NaT filter from the masked array tests (and moved it to the tests), this might also affect downstream. We could also call into the normal assert equal function (which would be cleaner), however currently that checks for scalar, which could be more disruptive then the possible NaT warning. Or we do that, and add a flag to assert equal to skip the scalar check

charris commented:
The masked testing utils are a bit of a mess anyway. In the future it would be nice if the testing functions were all clearly split between the usual numpy testing functions and those specific to masked arrays. I think what is done here is OK and we will deal with any problems that arise if/when they arise.

charris commented:
Thanks Sebastian.

seberg commented:
Thanks!

eric-wieser commented on numpy/testing/utils.py:
```diff
@@ -396,14 +397,24 @@ def assert_equal(actual,desired,err_msg='',verbose=True):
+        # If both are NaT (and have the same dtype -- datetime or timedelta)
+        # they are considered equal.
+        if (isnat(desired) == isnat(actual) and
+                array(desired).dtype.type == array(actual).dtype.type):
```
This isn't right - it considers any two non-NAT times equal!

seberg replied:
ooopst....

eric-wieser replied:
Fixed in #10096
